packages/core/test/SignUp.ts

```diff
@@ -80,5 +80,67 @@ describe('Signup', function () {
                 'ReachedMaximumNumberUserSignedUp'
             )
         })
+
+        it('setUsername should succeed', async () => {
+            const epoch = Number(await unirepContract.currentEpoch())
+            const epkNonce = 1
+            const randomEpochKey = genEpochKey(
+                genRandomSalt(),
+                epoch,
+                epkNonce
+            ).valueOf()
+            const oldUsername = 0
+            const newUsername = 1
+            const accounts = await ethers.getSigners()
+
+            const tx = await unirepSocialContract
+                .connect(accounts[0])
+                .setUsername(randomEpochKey, oldUsername, newUsername, {
+                    value: DEFAULT_ATTESTING_FEE,
+                })
+            const receipt = await tx.wait()
+
+            expect(receipt.status).equal(1)
+            const isClaimed = await unirepSocialContract.usernames(newUsername)
+
+            expect(isClaimed, 'This username has not been updated').to.be.true
+        })
+
+        it('should fail if a username is double registered', async () => {
+            const epoch = Number(await unirepContract.currentEpoch())
+            const epkNonce = 1
+            const randomEpochKey1 = genEpochKey(
+                genRandomSalt(),
+                epoch,
+                epkNonce
+            ).valueOf()
+            const randomEpochKey2 = genEpochKey(
+                genRandomSalt(),
+                epoch,
+                epkNonce
+            ).valueOf()
+            const oldUsername = 1
+            const newUsername = 2
+            const accounts = await ethers.getSigners()
+
+            // set newUserName to randomEpochKey1
+            await unirepSocialContract
+                .connect(accounts[0])
+                .setUsername(randomEpochKey1, oldUsername, newUsername, {
+                    value: DEFAULT_ATTESTING_FEE,
+                })
+                .then((t) => t.wait())
+
+            const oldUsername2 = 3
+
+            // try to set the same newUserName to randomEpochKey2
+            await expect(
+                unirepSocialContract
+                    .connect(accounts[0])
+                    .setUsername(randomEpochKey2, oldUsername2, newUsername, {
+                        value: DEFAULT_ATTESTING_FEE,
+                    })
+            ).to.be.revertedWith('This username is already taken')
+        })
```

Comment:
Other test cases:

- Should fail if called by non-admin
- Should mark old username as not claimed
- Should emit attestation event with graffiti matching the new username
  - [This](https://github.com/Unirep/Unirep/blob/main/packages/contracts/contracts/interfaces/IUnirep.sol#L22) is what the attestation event looks like
  - [This](https://hardhat.org/hardhat-runner/plugins/nomicfoundation-hardhat-chai-matchers#usage) is how you can test that a contract call emitted an event
  - Would be great if you could check the contents of the event to make sure it matches the username that was set. I'm not sure how to do it in the chai matchers though.
  - [This](https://github.com/Unirep/Unirep/blob/main/packages/core/src/Synchronizer.ts#L478) is an example decoding an attestation event, maybe it will help